### PR TITLE
Remember trusted senders

### LINKED ISSUE #505: Remember »Show images from this sender«
Clicking the button »Show images from this sender« should only be necessary once per sender. We should remember that you want to see images from this sender, which is the point of this button. ;)

cc @nextcloud/mail 

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/49057679-remember-show-images-from-this-sender?utm_campaign=plugin&utm_content=tracker%2F44154351&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F44154351&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by ChristophWurst**
Fixes #505

### CODE REVIEW

**Comment by GretaD** [comment]
@jancborchardt on the ticket  #505 you have written:
 _Clicking the button »Show images from this sender« should only be necessary once per sender._

But i guess you agree that if the person changes their mind then they should be able to undo the action, and hide images from a sender they marked as trusted to show images..?
In this case, how am I suppose to show this action (hide images) in the message? As a button, same as show images? Or do you have any better idea?
I checked Thunderbird and Gmail and i cannot find the option to hide the images after you click show.

**Comment by jancborchardt** [comment]
> I checked Thunderbird and Gmail and i cannot find the option to hide the images after you click show.

@GretaD yeah, that's why we also don't really need it. :)

**Comment by ChristophWurst** [comment]
> I checked Thunderbird and Gmail and i cannot find the option to hide the images after you click show.

See https://support.google.com/mail/thread/12526769?hl=en and

Bildschirmfoto von 2020-11-25 11-31-38

They both have it, and for a good reason. You can't otherwise undo this if you accidentally clicked the button. Not having this means we have to assume that users never make any mistakes and never revoke this. There could be someone who you formerly trust but then you maybe notice they send tracking pixels and you want to prevent them from invading your privacy.

**Comment by GretaD** [comment]
> > I checked Thunderbird and Gmail and i cannot find the option to hide the images after you click show.

> They both have it, and for a good reason. You can't otherwise undo this if you accidentally clicked the button. Not having this means we have to assume that users never make any mistakes and never revoke this. There could be someone who you formerly trust but then you maybe notice they send tracking pixels and you want to prevent them from invading your privacy.

Yeah, thanks Christoph, I said i couldnt find it not that they dont have it. I agree that we should not have irreversible actions on our app.

**Review by st3iny — APPROVED** [review]
Very nice! Just tested and it works well.

We should merge this and add an UI for revoking trust in a followup.

What about the PHP 7.2 lint though? Didn't we drop support for it?

**Comment by ChristophWurst** [comment]
> We should merge this and add an UI for revoking trust in a followup.

That is the plan :) @GretaD did you open a ticket?

> What about the PHP 7.2 lint though? Didn't we drop support for it?

Yeah, I hoped so. But unfortunately the server doesn't compare the specified php version, so installations of Nextcloud 20 got the 1.6 release and that broke HTML messages. See https://github.com/nextcloud/mail/pull/4150 for more details.

File: lib/Service/TrustedSenderService.php [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	private $mapper;
+
+	public function __construct(TrustedSenderMapper $mapper) {
+		$this->mapper = $mapper;
+	}
+
+	public function isTrusted(string $uid, string $email): bool {
+		return $this->mapper->exists(
+			$uid,
+			$email,
```
Comment (ChristophWurst) [inline lib/Service/TrustedSenderService.php:43]:
```suggestion
			$email
```

File: lib/Service/TrustedSenderService.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	public function trust(string $uid, string $email, ?bool $trust = true): void {
+		if ($trust && $this->isTrusted($uid, $email)) {
+			// Nothing to do
+			return;
+		}
+
+		if ($trust) {
+			$this->mapper->create(
+				$uid,
+				$email,
```
Comment (ChristophWurst) [inline lib/Service/TrustedSenderService.php:56]:
```suggestion
				$email
```

File: lib/Service/TrustedSenderService.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+			);
+		} else {
+			$this->mapper->remove(
+				$uid,
+				$email,
```
Comment (ChristophWurst) [inline lib/Service/TrustedSenderService.php:61]:
```suggestion
				$email
```

**Review by ChristophWurst — COMMENTED** [review]
To make the php7.2 linter happy

**Comment by GretaD** [comment]
> > We should merge this and add an UI for revoking trust in a followup.
> 
> That is the plan :) @GretaD did you open a ticket?
> 
Here  #4176, sorry i was suppose to open it yesterday :)

**Comment by NESC1US** [comment]
how about remember to not show any externally linking images from this (or any) sender? i dont want someone else being able to validate existence of my mail address, then spam me, and also put some cookies in.. thanks for thinking about the consequences of this "feature"

**Comment by st3iny** [comment]
@NESC1US By default **no** images in emails are shown/loaded. If you decide to trust a sender you can do so by clicking the trust sender button. See #4176 for revoking trust from a sender you previously trusted.

**Comment by NESC1US** [comment]
@st3iny cool, I will just use ublock origin to block that annoying notification.
this is already better than thunderbird where this workaround is not easily applicable.

**Comment by StCyr** [comment]
There's no notification: there's a button at the top of the message's body

**Comment by NESC1US** [comment]
> There's no notification: there's a button at the top of the message's body

whatever you want to call it, i dont like it being there in the first place, its setting up users for being pwned.

**Comment by ChristophWurst** [comment]
Hey @NESC1US I think there are a few misconceptions. Please look at the app again, like a current release. The button was there before. And even if you want to block all external resources generally, it's acceptable for some users to allow them.

This PR is done, there's nothing to discuss from here. If you have a bug report or feature request then please open one as new ticket.

And please stop by at https://nextcloud.com/contribute/code-of-conduct/.

**Pull request merged**